sdk/nodejs/runtime/closure/codePaths.ts

```diff
@@ -0,0 +1,186 @@
+    extraIncludePackages = extraIncludePackages || [];
+    extraExcludePackages = extraExcludePackages || [];
+
+    const includedPackages = new Set<string>();
+    const excludedPackages = new Set<string>();
+
+    for (const p of extraExcludePackages) {
+        excludedPackages.add(p);
+    }
+
+    for (const p of extraIncludePackages) {
+        includedPackages.add(p);
+    }
+
+    // Find folders for all packages requested by the user
+    const pathSet = await allFoldersForPackages(includedPackages, excludedPackages);
+
+    // Add all paths explicitly requested by the user
+    for (const path of extraIncludePaths) {
+        pathSet.add(path);
+    }
+
+    const codePaths: Map<string, asset.Asset | asset.Archive> = new Map();
+
+    // For each of the required paths, add the corresponding FileArchive or FileAsset to the AssetMap.
+    for (const path of pathSet.values()) {
+        // The Asset model does not support a consistent way to embed a file-or-directory into an `AssetArchive`, so
+        // we stat the path to figure out which it is and use the appropriate Asset constructor.
+        const stats = fs.lstatSync(path);
+        if (stats.isDirectory()) {
+            codePaths.set(path, new asset.FileArchive(path));
+        }
+        else {
+            codePaths.set(path, new asset.FileAsset(path));
+        }
+    }
+
+    return codePaths;
+}
+
+const lambdaRolePolicy = {
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Action": "sts:AssumeRole",
+            "Principal": {
+                "Service": "lambda.amazonaws.com",
+            },
+            "Effect": "Allow",
+            "Sid": "",
+        },
+    ],
+};
+
+// Package is a node in the package tree returned by readPackageTree.
+interface Package {
+    name: string;
+    path: string;
+    package: {
+        dependencies?: { [key: string]: string; };
+    };
+    parent?: Package;
+    children: Package[];
+    error: any;
+}
+
+// allFolders computes the set of package folders that are transitively required by the root
+// 'dependencies' node in the client's project.json file.
+function allFoldersForPackages(includedPackages: Set<string>, excludedPackages: Set<string>): Promise<Set<string>> {
+    return new Promise((resolve, reject) => {
+        readPackageTree(".", undefined, (err: any, root: Package) => {
+            if (err) {
+                return reject(err);
+            }
+
+            // read-package-tree defers to read-package-json to parse the project.json file. If that
+            // fails, root.error is set to the underlying error.  In that case, we want to fail as
+            // well.  Otherwise, we will silently proceed as if package.json was empty, which would
+            // result in us uploading no node_modules.
+            if (root.error) {
+                return reject(new RunError(
+                    "Failed to parse package.json. Underlying issue:\n  " + root.error.toString()));
+            }
```

Comment:
this is a new addition.  It is to fix: https://github.com/pulumi/pulumi/issues/1702